### PR TITLE
Modernized terraform script for deployment as a module of nasa-pds/registry

### LINKED ISSUE #4: cicd fails on registry-api-service
image

See for detals action https://github.com/NASA-PDS/registry-api-service/actions/runs/1575429493

### LINKED ISSUE #767: When a request /classes/{product class} does not match an existing product class, I want a 404 error.
### Checked for duplicates

No - I haven't checked

### 🐛 Describe the bug

When I did /classes/collections action, I got a 400 error.

### 🕵️ Expected behavior

The integration tests expect a 404 error.

### 📜 To Reproduce

1.
2.
3.
...

### 🖥 Environment Info

- Version of this software [e.g. vX.Y.Z]
- Operating System: [e.g. MacOSX with Docker Desktop vX.Y]
...

### 📚 Version of Software Used

_No response_

### 🩺 Test Data / Additional context

_No response_

### 🦄 Related requirements

🦄 #xyz

### ⚙️ Engineering Details

_No response_

### 🎉 Integration & Test

_No response_

### PR DESCRIPTION
**Pull request by tloubrieu-jpl**
## 🗒️ Summary

Modernize the terraform code for consistency with other repository and so to be able to call the code as a module of the umbrella repository nasa-pds/registry deployment script.

Aborted attempt to use Pull Through Cache ECR configuration instead of a manual deployment of the docker image. We will have to come back to that later since it did not work.

### 🤖 AI Assistance Disclosure
- [ ] No AI assistance used
- [ ] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)
- [ ] AI used for moderate content generation (AI generated some code or logic, but the developer authored or heavily revised the majority)
- [x] AI generated substantial portions of this code

Estimated % of code influenced by AI: __50_ %

## ⚙️ Test Data and/or Report

Tested with a deployment on AWS and integration tests succesfully run on it.

## ♻️ Related Issues

Also fixes https://github.com/NASA-PDS/registry-api/issues/767 (integration test fix)

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### CODE REVIEW

File: terraform/main.tf [hunk 1]
```diff
@@ -75,44 +75,64 @@ resource "aws_lb_listener_rule" "pds-registry-forward-rule" {
...
+  count = 1 - var.create_github_secret_credentials
+  name  = "ecr-pullthroughcache/github-credentials"
+}
+
+locals {
+  github_ecr_credentials_arn = var.create_github_secret_credentials == 1 ? aws_secretsmanager_secret.github_ecr_credentials[0].arn : data.aws_secretsmanager_secret.github_ecr_credentials[0].arn
+}
+
+# Add a Pull Through Cache rule for GHCR
+resource "aws_ecr_pull_through_cache_rule" "ghcr" {
```
Comment (nutjob4life) [inline terraform/main.tf:108]: Wow this is really cool. I had no idea this resource even existed.

File: .gitignore [hunk 2]
```diff
@@ -65,6 +65,9 @@ src/test/temp/
 
 # terraform
 .terraform/
+terraform/*.tfvars
+!terraform/*.tfvars.example
+terraform/.terraform.lock.hcl
```
Comment (nutjob4life) [inline .gitignore]: I believe the `.terraform.lock.hcl` has the same purpose as the `package-lock.json` in Node projects and exists to pin provider versions, avoiding the "works on my machine" issue. AI agrees and says `.terraform.lock.hcl` should be commited and the `.gitingore` line should be

    !.terraform.lock.hcl

File: terraform/main.tf [hunk 3]
```diff
@@ -75,44 +75,64 @@ resource "aws_lb_listener_rule" "pds-registry-forward-rule" {
...
+  secret_id     = aws_secretsmanager_secret.github_ecr_credentials[count.index].id
+  secret_string = jsonencode({
+    username = var.github_username
+    accessToken    = var.github_token
+  })
+}
+
+# Look up the secret when it is not created by this script
+data "aws_secretsmanager_secret" "github_ecr_credentials" {
+  count = 1 - var.create_github_secret_credentials
```
Comment (nutjob4life) [inline terraform/main.tf:99]: Might be safer to to do
```hcl
type = bool
default = true
```
and do
```hcl
count = var.create_github_secret_credentials ? 1 : 0
```
on the off chance that someone sets `create_github_secret_credentials = 2`, for example.

File: terraform/main.tf [hunk 4]
```diff
@@ -75,44 +75,64 @@ resource "aws_lb_listener_rule" "pds-registry-forward-rule" {
...
 
-# Do we need individual dev/test/prod repositories?
-# I don't think we do, but then we need to use prod account instead of the dev account, would that work ?
-data "aws_ecr_repository" "pds-registry-api-service" {
-  name = "pds-registry-api-service"
+resource "aws_secretsmanager_secret_version" "github_ecr_credentials" {
+  count = var.create_github_secret_credentials
+
+  secret_id     = aws_secretsmanager_secret.github_ecr_credentials[count.index].id
+  secret_string = jsonencode({
```
Comment (nutjob4life) [inline terraform/main.tf:91]: Might want to add a warning note to the README that the GitHub username and token become part of the Terraform state.

File: terraform/output.tf [hunk 5]
```diff
@@ -0,0 +1,19 @@
+locals {
+
+  module_relative_path = replace(abspath(path.module), "/^.*\\/terraform(\\/|$)/", "")
```
Comment (nutjob4life) [inline terraform/output.tf:3]: Does this work?  I didn't think Terraform's `replace` uses `/.../` delimiters for regexes.

**Review by nutjob4life — CHANGES_REQUESTED** [review]
Excellent modernization of the Terraform structure that should make it more maintanable and easier to use. Just a few nits (see inline comments). Thanks!

File: terraform/output.tf (hunk 5 above)
Comment (tloubrieu-jpl) [reply terraform/output.tf:3]: claude code came up with that syntax and it works.

File: terraform/main.tf (hunk 1 above)
Comment (tloubrieu-jpl) [reply terraform/main.tf:108]: Unfortunatly I was unable to make it work with a terrafrom deployment, I'll have to come back to it later.

**Comment by jordanpadams** [comment]
@tloubrieu-jpl when you have a chance can you respond to all of @nutjob4life comments and triage the sonarcloud issues?

**Review by nutjob4life — APPROVED** [review]
Thanks for the updates @tloubrieu-jpl 👍

**Comment by jordanpadams** [comment]
Numerous things called out by sonarcloud but will merge for now. Need to move this forward.

**Review by jordanpadams — APPROVED** [review]
